src/one_any_of.rs

```diff
@@ -0,0 +1,168 @@
+                match self {
+                    $(Self::$i(inner) => inner.serialize(serializer)),*
+                }
+            }
+        }
+
+        impl<$($i),*> ToString for $t<$($i),*> where
+            $($i: PartialEq + ToString,)*
+        {
+            fn to_string(&self) -> String {
+                match self {
+                    $(Self::$i(inner) => inner.to_string()),*
+                }
+            }
+        }
+    }
+}
+
+// Define a macro to define the `OneOf` enum for a specific number of inner types.
+macro_rules! one_of {
+    (
+        $t:ident,
+        $($i:ident),*
+    ) => {
+        common_one_any_of!($t, $($i),*);
+
+        impl<'b, $($i),*> Deserialize<'b> for $t<$($i),*> where
+            $($i: PartialEq + for<'a> Deserialize<'a>,)*
+        {
+            fn deserialize<De: Deserializer<'b>>(deserializer: De) -> Result<Self, De::Error> {
+                let content = Content::deserialize(deserializer)?;
+                let mut result = Err(De::Error::custom("data did not match any within oneOf"));
+                $(
+                    if let Ok(inner) = $i::deserialize(ContentRefDeserializer::<De::Error>::new(&content)) {
+                        if result.is_err() {
+                            result = Ok(Self::$i(inner));
+                        } else {
+                            return Err(De::Error::custom("data matched multiple within oneOf"))
+                        }
+                    }
+                )*
+                result
+            }
+        }
+
+        impl<$($i),*> FromStr for $t<$($i),*> where
+            $($i: PartialEq + FromStr,)*
+        {
+            type Err = &'static str;
+            fn from_str(x: &str) -> Result<Self, Self::Err> {
+                let mut result = Err("data did not match any within oneOf");
+                $(
+                    if let Ok(inner) = $i::from_str(x) {
+                        if result.is_err() {
+                            result = Ok(Self::$i(inner));
+                        } else {
+                            return Err("data matched multiple within oneOf")
+                        }
+                    }
+                )*
+                result
+            }
+        }
+    }
+}
+
+// Use the `one_of!` macro to define the `OneOf` enum for 1-16 inner types.
+one_of!(OneOf1, A);
+one_of!(OneOf2, A, B);
+one_of!(OneOf3, A, B, C);
+one_of!(OneOf4, A, B, C, D);
+one_of!(OneOf5, A, B, C, D, E);
+one_of!(OneOf6, A, B, C, D, E, F);
+one_of!(OneOf7, A, B, C, D, E, F, G);
+one_of!(OneOf8, A, B, C, D, E, F, G, H);
+one_of!(OneOf9, A, B, C, D, E, F, G, H, I);
+one_of!(OneOf10, A, B, C, D, E, F, G, H, I, J);
+one_of!(OneOf12, A, B, C, D, E, F, G, H, I, J, K);
+one_of!(OneOf13, A, B, C, D, E, F, G, H, I, J, K, L);
+one_of!(OneOf14, A, B, C, D, E, F, G, H, I, J, K, L, M);
+one_of!(OneOf15, A, B, C, D, E, F, G, H, I, J, K, L, M, N);
+one_of!(OneOf16, A, B, C, D, E, F, G, H, I, J, K, L, M, N, O);
```

Comment:
Instead of having numbered types, did you consider doing:

```
use serde::{
    de::Error,
    private::de::{Content, ContentRefDeserializer},
    Deserialize, Deserializer, Serialize, Serializer,
};

// As per https://github.com/rust-lang/rust/issues/34537
mod hidden {
   pub struct Private;
}

use hidden::Private;

impl<'de> Deserialize<'de> for Private {
    fn deserialize<De: Deserializer<'de>>(deserializer: De) -> Result<Self, De::Error> {
        return Err(De::Error::custom("invalid option"))
    }
}

pub enum OneOf<A,
           B=Private,
           C=Private,
           D=Private,
           E=Private,
           F=Private,
           G=Private,
           H=Private,
           I=Private,
           J=Private,
           K=Private,
           L=Private,
           M=Private,
           N=Private,
           O=Private> {
    A(A),
    B(B),
    C(C),
    D(D),
    E(E),
    F(F),
    G(G),
    H(H),
    I(I),
    J(J),
    K(K),
    L(L),
    M(M),
    N(N),
    O(O)
}

fn main() {
    let foo = OneOf::<u8, u16>::B(5);
}
```

Reply:
I guess this makes matching worse, as you have to match on a bunch of options which are unhittable...

Reply:
No, happy to switch to this if you prefer - let me know!

Reply:
Given that it'll force non-exhaustive matching, it's probably better as is.